File: backend/app/schemas/business_os.py
"""Pydantic V2 Schemas for Business OS, Strategy, KPIs, Decisions, and Executive Intelligence."""

from datetime import datetime, timezone
from decimal import Decimal
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field

from app.business_os.base import (
    AlertSeverity,
    AlertStatus,
    BriefingFrequency,
    BusinessHealthStatus,
    DecisionPriority,
    DecisionStatus,
    InitiativeStatus,
    KPICategory,
    KeyResultStatus,
    ObjectivePriority,
    ObjectiveStatus,
    RiskCategory,
    RiskImpact,
    RiskLifecycleStatus,
    RiskProbability,
    RiskSeverity,
    ScenarioType,
    ScorecardStatus,
    StrategicDependencyState,
)


# --- Strategy & OKR Schemas ---

class KeyResultCreate(BaseModel):
    title: str
    target_value: Decimal
    current_value: Decimal = Decimal("0.00")
    unit: str = "%"
    owner: Optional[str] = None


class KeyResultResponse(BaseModel):
    kr_id: str
    objective_id: str
    title: str
    target_value: Decimal
    current_value: Decimal
    unit: str
    status: KeyResultStatus
    progress_pct: Decimal
    owner: Optional[str] = None


class InitiativeCreate(BaseModel):
    title: str
    owner: str
    budget: Decimal = Decimal("0.00")
    priority: ObjectivePriority = ObjectivePriority.P1_HIGH


class InitiativeResponse(BaseModel):
    initiative_id: str
    objective_id: str
    title: str
    owner: str
    budget: Decimal
    priority: ObjectivePriority
    status: InitiativeStatus
    progress_pct: Decimal
    milestones: List[Dict[str, Any]] = Field(default_factory=list)
    dependencies: List[Dict[str, Any]] = Field(default_factory=list)


class StrategicObjectiveCreate(BaseModel):
    title: str
    description: Optional[str] = None
    timeframe: str = "FY2026"
    priority: ObjectivePriority = ObjectivePriority.P1_HIGH
    owner: str = "Executive Team"


class StrategicObjectiveResponse(BaseModel):
    objective_id: str
    title: str
    description: Optional[str] = None
    timeframe: str
    priority: ObjectivePriority
    status: ObjectiveStatus
    owner: str
    overall_progress_pct: Decimal
    key_results: List[KeyResultResponse] = Field(default_factory=list)
    initiatives: List[InitiativeResponse] = Field(default_factory=list)


# --- KPI & Scorecard Schemas ---

class KPIDefinitionCreate(BaseModel):
    kpi_id: str
    name: str
    category: KPICategory
    description: str
    unit: str = ""
    currency: str = "PKR"
    source_domain: str
    formula: str
    target_value: Optional[Decimal] = None
    warning_threshold: Optional[Decimal] = None
    critical_threshold: Optional[Decimal] = None
    freshness_max_seconds: int = 3600
    is_higher_better: bool = True


class KPISnapshotResponse(BaseModel):
    kpi_id: str
    name: str
    category: KPICategory
    value: Decimal
    previous_value: Optional[Decimal] = None
    target_value: Optional[Decimal] = None
    variance: Optional[Decimal] = None
    variance_pct: Optional[Decimal] = None
    unit: str
    currency: str
    source_domain: str
    version: str
    freshness_seconds: int
    is_stale: bool
    status: ScorecardStatus
    calculated_at: datetime


class ScorecardItemResponse(BaseModel):
    kpi_id: str
    name: str
    category: KPICategory
    target: Decimal
    actual: Decimal
    variance: Decimal
    variance_pct: Optional[Decimal]
    unit: str
    currency: str
    trend: str
    status: ScorecardStatus
    owner: str


class DepartmentScorecardResponse(BaseModel):
    department_name: str
    category: KPICategory
    overall_status: ScorecardStatus
    scorecard_items: List[ScorecardItemResponse]
    generated_at: datetime


class CrossDomainReconciliationRequest(BaseModel):
    finance_revenue: Decimal
    crm_closed_won_revenue: Decimal
    active_contracts_count: int
    active_billing_profiles_count: int


# --- Business Health Schemas ---

class HealthDimensionScoreResponse(BaseModel):
    dimension_name: str
    score: Decimal
    weight_pct: Decimal
    status: BusinessHealthStatus
    positive_drivers: List[str]
    negative_drivers: List[str]
    signals_count: int


class BusinessHealthReportResponse(BaseModel):
    overall_health_score: Decimal
    overall_status: BusinessHealthStatus
    dimensions: Dict[str, HealthDimensionScoreResponse]
    key_strengths: List[str]
    critical_risks: List[str]
    reconciliation_alerts: List[str]
    evaluated_at: datetime


# --- Risk Register Schemas ---

class RiskCreate(BaseModel):
    title: str
    description: str
    category: RiskCategory
    probability: RiskProbability = RiskProbability.POSSIBLE
    impact: RiskImpact = RiskImpact.MODERATE
    owner: str
    evidence_signals: List[str] = Field(default_factory=list)
    mitigation_strategy: str = ""
    contingency_plan: str = ""
    due_date: Optional[str] = None


class RiskResponse(BaseModel):
    risk_id: str
    title: str
    description: str
    category: RiskCategory
    probability: RiskProbability
    impact: RiskImpact
    risk_score: int
    severity: RiskSeverity
    owner: str
    status: RiskLifecycleStatus
    identified_at: datetime
    due_date: Optional[str] = None
    evidence_signals: List[str]
    mitigation_strategy: str
    contingency_plan: str
    last_reviewed_at: datetime


# --- Decision Queue Schemas ---

class DecisionCreate(BaseModel):
    title: str
    business_question: str
    context_summary: str
    priority: DecisionPriority = DecisionPriority.P1_HIGH
    evidence_signals: List[str] = Field(default_factory=list)
    expected_outcome: str = ""
    review_due_date: Optional[str] = None


class DecisionRecordAction(BaseModel):
    chosen_option_id: str
    decision_rationale: str
    decided_by: str


class DecisionOutcomeRecordAction(BaseModel):
    actual_outcome: str
    outcome_variance_analysis: str
    lessons_learned: List[str]


class DecisionResponse(BaseModel):
    decision_id: str
    title: str
    business_question: str
    context_summary: str
    priority: DecisionPriority
    status: DecisionStatus
    candidate_options: List[Dict[str, Any]]
    ai_recommendation: Optional[str] = None
    ai_recommendation_rationale: Optional[str] = None
    chosen_option_id: Optional[str] = None
    chosen_option_title: Optional[str] = None
    decision_rationale: Optional[str] = None
    decided_by: Optional[str] = None
    decided_at: Optional[datetime] = None
    evidence_signals: List[str]
    expected_outcome: str
    actual_outcome: Optional[str] = None
    outcome_variance_analysis: Optional[str] = None
    lessons_learned: List[str]
    created_at: datetime
    review_due_date: Optional[str] = None


# --- Scenario Planning Schemas ---

class ScenarioRunRequest(BaseModel):
    scenario_name: str
    scenario_type: ScenarioType = ScenarioType.CUSTOM
    base_revenue: Decimal = Decimal("4000000.00")
    base_cost: Decimal = Decimal("1400000.00")
    base_capacity_hours: Decimal = Decimal("160.0")
    base_committed_hours: Decimal = Decimal("140.0")
    price_change_pct: Decimal = Decimal("0.0")
    conversion_change_pct: Decimal = Decimal("0.0")
    client_churn_revenue: Decimal = Decimal("0.0")
    new_hires_count: int = 0
    additional_project_hours: Decimal = Decimal("0.0")


class ScenarioResponse(BaseModel):
    scenario_id: str
    scenario_name: str
    scenario_type: ScenarioType
    simulated_revenue: Decimal
    simulated_profit: Decimal
    simulated_margin_pct: Decimal
    capacity_utilization_pct: Decimal
    cash_requirement: Decimal
    risk_level: str
    assumptions_applied: Dict[str, Any]
    sensitivity_rankings: List[Dict[str, Any]]
    generated_at: datetime
    is_production_isolated: bool


# --- Briefings, Calendar & Copilot Schemas ---

class ExecutiveBriefingResponse(BaseModel):
    briefing_id: str
    frequency: BriefingFrequency
    title: str
    summary_paragraph: str
    key_metrics_snapshot: Dict[str, Any]
    what_changed_summary: List[str]
    top_decisions_required: List[Dict[str, Any]]
    critical_risks: List[Dict[str, Any]]
    upcoming_deadlines: List[Dict[str, Any]]
    recommended_attention_areas: List[str]
    generated_at: datetime


class BusinessCalendarEventResponse(BaseModel):
    event_id: str
    title: str
    event_type: str
    event_date: str
    related_entity_id: str
    related_entity_name: str
    severity: str
    owner: str
    is_completed: bool


class CopilotQueryRequest(BaseModel):
    query: str
    user_role: str = "EXECUTIVE"


class CopilotQueryResponse(BaseModel):
    query_id: str
    user_query: str
    intent: str
    answer_markdown: str
    confidence: str
    evidence_sources: List[Dict[str, Any]]
    suggested_followups: List[str]
    action_prohibited: bool
    prohibited_reason: Optional[str] = None
    answered_at: datetime

File: frontend/lib/api/business_os.ts
import { api } from './client';

export interface KPISnapshot {
  kpi_id: string;
  name: string;
  category: string;
  value: number;
  previous_value?: number;
  target_value?: number;
  variance?: number;
  variance_pct?: number;
  unit: string;
  currency: string;
  source_domain: string;
  version: string;
  freshness_seconds: number;
  is_stale: boolean;
  status: 'EXCEEDING' | 'ON_TRACK' | 'AT_RISK' | 'OFF_TRACK';
  calculated_at: string;
}

export interface HealthDimension {
  dimension_name: string;
  score: number;
  weight_pct: number;
  status: 'HEALTHY' | 'STABLE' | 'WATCH' | 'AT_RISK' | 'CRITICAL';
  positive_drivers: string[];
  negative_drivers: string[];
  signals_count: number;
}

export interface BusinessHealthReport {
  overall_health_score: number;
  overall_status: 'HEALTHY' | 'STABLE' | 'WATCH' | 'AT_RISK' | 'CRITICAL';
  dimensions: Record<string, HealthDimension>;
  key_strengths: string[];
  critical_risks: string[];
  reconciliation_alerts: string[];
  evaluated_at: string;
}

export interface StrategicKeyResult {
  kr_id: string;
  objective_id: string;
  title: string;
  target_value: number;
  current_value: number;
  unit: string;
  status: string;
  progress_pct: number;
  owner?: string;
}

export interface StrategicInitiative {
  initiative_id: string;
  objective_id: string;
  title: string;
  owner: string;
  budget: number;
  priority: string;
  status: string;
  progress_pct: number;
  milestones: any[];
  dependencies: any[];
}

export interface StrategicObjective {
  objective_id: string;
  title: string;
  description?: string;
  timeframe: string;
  priority: string;
  status: string;
  owner: string;
  overall_progress_pct: number;
  key_results: StrategicKeyResult[];
  initiatives: StrategicInitiative[];
}

export interface DepartmentScorecard {
  department_name: string;
  category: string;
  overall_status: string;
  scorecard_items: Array<{
    kpi_id: string;
    name: string;
    category: string;
    target: number;
    actual: number;
    variance: number;
    variance_pct?: number;
    unit: string;
    currency: string;
    trend: string;
    status: string;
    owner: string;
  }>;
  generated_at: string;
}

export interface OrganizationalRisk {
  risk_id: string;
  title: string;
  description: string;
  category: string;
  probability: string;
  impact: string;
  risk_score: number;
  severity: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  owner: string;
  status: string;
  identified_at: string;
  due_date?: string;
  evidence_signals: string[];
  mitigation_strategy: string;
  contingency_plan: string;
  last_reviewed_at: string;
}

export interface DecisionOption {
  option_id: string;
  title: string;
  description: string;
  expected_benefits: string[];
  expected_costs: string[];
  risks: string[];
  dependencies: string[];
  estimated_impact: string;
  confidence: string;
  evidence_references: string[];
}

export interface DecisionRecord {
  decision_id: string;
  title: string;
  business_question: string;
  context_summary: string;
  priority: string;
  status: string;
  candidate_options: DecisionOption[];
  ai_recommendation?: string;
  ai_recommendation_rationale?: string;
  chosen_option_id?: string;
  chosen_option_title?: string;
  decision_rationale?: string;
  decided_by?: string;
  decided_at?: string;
  evidence_signals: string[];
  expected_outcome: string;
  actual_outcome?: string;
  outcome_variance_analysis?: string;
  lessons_learned: string[];
  created_at: string;
  review_due_date?: string;
}

export interface ScenarioSimulationResult {
  scenario_id: string;
  scenario_name: string;
  scenario_type: string;
  simulated_revenue: number;
  simulated_profit: number;
  simulated_margin_pct: number;
  capacity_utilization_pct: number;
  cash_requirement: number;
  risk_level: string;
  assumptions_applied: Record<string, any>;
  sensitivity_rankings: Array<{
    variable: string;
    sensitivity_level: string;
    impact_description: string;
    elasticity_score: number;
  }>;
  generated_at: string;
  is_production_isolated: boolean;
}

export interface ExecutiveBriefing {
  briefing_id: string;
  frequency: string;
  title: string;
  summary_paragraph: string;
  key_metrics_snapshot: Record<string, any>;
  what_changed_summary: string[];
  top_decisions_required: any[];
  critical_risks: any[];
  upcoming_deadlines: any[];
  recommended_attention_areas: string[];
  generated_at: string;
}

export interface BusinessCalendarEvent {
  event_id: string;
  title: string;
  event_type: string;
  event_date: string;
  related_entity_id: string;
  related_entity_name: string;
  severity: string;
  owner: string;
  is_completed: boolean;
}

export interface CopilotResponse {
  query_id: string;
  user_query: string;
  intent: string;
  answer_markdown: string;
  confidence: string;
  evidence_sources: any[];
  suggested_followups: string[];
  action_prohibited: boolean;
  prohibited_reason?: string;
  answered_at: string;
}

export const businessOSApi = {
  getOverview: async () => {
    return api.get<any>('/executive/overview');
  },
  getHealth: async () => {
    return api.get<BusinessHealthReport>('/executive/health');
  },
  getBriefing: async (frequency: string = 'DAILY') => {
    return api.get<ExecutiveBriefing>(`/executive/briefings?frequency=${frequency}`);
  },
  getCalendar: async () => {
    return api.get<BusinessCalendarEvent[]>('/executive/calendar');
  },
  queryCopilot: async (query: string, user_role: string = 'EXECUTIVE') => {
    return api.post<CopilotResponse>('/executive/assistant/query', { query, user_role });
  },
  listObjectives: async () => {
    return api.get<StrategicObjective[]>('/strategy/objectives');
  },
  listScorecards: async () => {
    return api.get<DepartmentScorecard[]>('/strategy/scorecards');
  },
  listKPIs: async (category?: string) => {
    const url = category ? `/kpis?category=${category}` : '/kpis';
    return api.get<KPISnapshot[]>(url);
  },
  listRisks: async (category?: string, minSeverity?: string) => {
    let url = '/risks?';
    if (category) url += `category=${category}&`;
    if (minSeverity) url += `min_severity=${minSeverity}&`;
    return api.get<OrganizationalRisk[]>(url);
  },
  listDecisions: async (status?: string) => {
    const url = status ? `/decisions?status=${status}` : '/decisions';
    return api.get<DecisionRecord[]>(url);
  },
  recordDecision: async (decisionId: string, chosenOptionId: string, rationale: string, decidedBy: string) => {
    return api.post<DecisionRecord>(`/decisions/${decisionId}/decide`, {
      chosen_option_id: chosenOptionId,
      decision_rationale: rationale,
      decided_by: decidedBy,
    });
  },
  recordDecisionOutcome: async (decisionId: string, actualOutcome: string, varianceAnalysis: string, lessons: string[]) => {
    return api.post<DecisionRecord>(`/decisions/${decisionId}/outcome`, {
      actual_outcome: actualOutcome,
      outcome_variance_analysis: varianceAnalysis,
      lessons_learned: lessons,
    });
  },
  runScenario: async (payload: any) => {
    return api.post<ScenarioSimulationResult>('/scenarios/run', payload);
  },
};
